src/node/ServiceEndpoint.js

```diff
@@ -0,0 +1,163 @@
+         * @description The port of the service endpoint.
+         */
+        this._port = props?.port != null ? props.port : null;
+
+        /**
+         * @type {?string}
+         * @description A node domain name. This MUST be the
+         * fully qualified domain(DNS) name of the node. This
+         * value MUST NOT be more than 253 characters.
+         * domain_name and ipAddressV4 are mutually exclusive.
+         * When the `domain_name` field is set, the `ipAddressV4`
+         * field MUST NOT be set. When the `ipAddressV4` field
+         * is set, the `domain_name` field MUST NOT be set.
+         */
+        this._domainName = props?.domainName != null ? props.domainName : null;
+    }
+
+    /**
+     * @param {Uint8Array} ipAddressV4
+     * @description Set 4-byte IPv4 address of the endpoint.
+     * @returns {ServiceEndpoint}
+     *
+     */
+    setIpAddressV4(ipAddressV4) {
+        if (this._domainName != null) {
+            throw new Error(
+                "Cannot set IP address when domain name is already set.",
+            );
+        }
+        this._ipAddressV4 = ipAddressV4;
+        return this;
+    }
+
+    /**
+     * @description Get 4-byte IPv4 address of the endpoint.
+     * @returns {?Uint8Array}
+     *
+     */
+    get getIpAddressV4() {
+        return this._ipAddressV4;
+    }
+
+    /**
+     * @param {number} port
+     * @description Set port of the endpoint.
+     * @returns {ServiceEndpoint}
+     *
+     */
+    setPort(port) {
+        this._port = port;
+        return this;
+    }
+
+    /**
+     * @description Get port of the endpoint.
+     * @returns {?number}
+     *
+     */
+    get getPort() {
+        return this._port;
+    }
+
+    /**
+     * @param {string} domainName
+     * @description Set domain name of the endpoint.
+     * @returns {ServiceEndpoint}
+     *
+     */
+    setDomainName(domainName) {
+        if (this._ipAddressV4 != null) {
+            throw new Error(
+                "Cannot set domain name when IP address is already set.",
+            );
+        }
+        this._domainName = domainName;
+        return this;
+    }
+
+    /**
+     * @description Get domain name of the endpoint.
+     * @returns {?string}
+     *
+     */
+    get getDomainName() {
+        return this._domainName;
+    }
+
+    /**
+     * @internal
+     * @param {IServiceEndpoint} serviceEndpoint
+     * @returns {ServiceEndpoint}
+     */
+    static _fromProtobuf(serviceEndpoint) {
+        return new ServiceEndpoint({
+            ipAddressV4:
+                serviceEndpoint.ipAddressV4 != null
+                    ? serviceEndpoint.ipAddressV4
+                    : undefined,
+            port:
+                serviceEndpoint.port != null ? serviceEndpoint.port : undefined,
+            domainName:
+                serviceEndpoint.domainName != null
+                    ? serviceEndpoint.domainName
+                    : undefined,
+        });
```

Comment:
I think some type of error handling can be implemented here. For example we can check whether ipAddressV4 and port are different than null or undefined AND domainName is null and vice versa. And after all the checks are passed to `return new ServiceEndpoint`. I suggest that because imagine a case where the three of `ipAddressV4`, `port` and `domainName` or null or undefined. Is this scenario even possible? @svetoslav-nikol0v

Reply:
So you mean we need to add a check if `ipAddressV4` or `domainName` is set and another check if the `port` is also set, right?